packages/solo/src/chain-cosmos-sdk.js

```diff
@@ -381,137 +330,210 @@ ${chainID} chain does not yet know of address ${myAddr}${adviseEgress(myAddr)}
           const out = JSON.parse(stdout);
-          if (Number(out.height) > 0) {
-            // We submitted the transaction successfully.
-            return {};
-          }
-          assert.fail(X`Unexpected output: ${stdout.trimRight()}`);
-        },
-        undefined,
-        {}, // defaultIfCancelled
-      );
-      return qret;
+
+          assert.equal(
+            parseInt(out.code, 10),
+            0,
+            X`Unexpected output: ${stdout.trimRight()}`,
+          );
+
+          // We submitted the transaction successfully.
+          sequenceNumber += 1n;
+        }
+      }
+    } catch (e) {
+      // Put back the deliveries we tried to make.
+      messagePool = messagePool.concat(messages);
+      throw e;
     } finally {
       if (tmpInfo) {
         await fs.promises.unlink(tmpInfo.path);
       }
     }
+  };
+
+  /**
+   * This function is entered at most the same number of times
+   * as the blockNotifier announces a new block.
+   *
+   * It then gets the mailbox.  There are no optimisations.
+   *
+   * @param {number=} lastBlockUpdate
+   */
+  const recurseEachNewBlock = async (lastBlockUpdate = undefined) => {
+    const { updateCount, value } = await blockNotifier.getUpdateSince(
+      lastBlockUpdate,
+    );
+    assert(value, X`${GCI} unexpectedly finished!`);
+    console.debug(`new block on ${GCI}, fetching mailbox`);
+    await getMailbox()
+      .then(ret => {
+        if (!ret) {
+          return;
+        }
+        const { outbox, ack } = ret;
+        // console.debug('have outbox', outbox, ack);
+        inbound(GCI, outbox, ack);
+      })
+      .catch(e => console.error(`Failed to fetch ${GCI} mailbox:`, e));
+    recurseEachNewBlock(updateCount);
+  };
+
+  // Begin the block consumer.
+  recurseEachNewBlock();
+
+  // Retry sending, but no more than one pending retry at a time.
+  let retryPending;
+  const retrySend = (e = undefined) => {
+    if (e) {
+      console.error(`Error sending`, e);
+    }
+    if (retryPending !== undefined) {
+      return;
+    }
+
+    retryPending = setTimeout(() => {
+      retryPending = undefined;
+      sendUpdater.updateState(true);
+    }, SEND_RETRY_DELAY_MS);
+  };
+
+  // This function ensures we only have one outgoing send operation at a time.
+  const recurseEachSend = async (lastSendUpdate = undefined) => {
+    // See when there is another requested send since our last time.
+    const { updateCount, value } = await sendNotifier.getUpdateSince(
+      lastSendUpdate,
+    );
+    assert(value, X`Sending unexpectedly finished!`);
+
+    await sendFromMessagePool().catch(retrySend);
+    recurseEachSend(updateCount);
+  };
+
+  // Begin the sender.
+  recurseEachSend();
+
+  async function deliver(newMessages, acknum) {
+    let doSend = false;
+    if (acknum > highestAck) {
+      highestAck = acknum;
+      doSend = true;
+    }
+    if (newMessages.length) {
+      messagePool = messagePool.concat(newMessages);
```

Comment:
We should look at the way the mailbox-scanning code calls `deliver()` to see if it repeats messages that have not yet been acked. If so, this `concat` should become more of an ordered-Set merge.